### PR TITLE
[kube-prometheus-stack] multi tenant alert routing

### LINKED ISSUE #1231: kube-state alert routing in multi tenant cluster
Hey guys,

I'm facing the following challenge. We are running a cluster with multiple tenant supporting their own workloads. Each tenant has a separate alerting channel, so we would need to route the alerts generated out of the kube-state-metrics based on the source namespaces.  As far as I can tell, the current chart does not allow such a dynamic routing of the alerts based on the source namespace.

We came up with the following solution:
1. add a label `alerts` to each namespace representing the name of the channel where the alert are supposed to be routed to.
```
kind: Namespace
metadata:
  labels:
    alerts: channel
```
2. leveraging the `kube_namespace_labels` metric, which contains all the labels attached to namespaces ( see `label_alerts` )
```
kube_namespace_labels{container="kube-state-metrics", endpoint="http", instance="10.32.3.33:8080", job="kube-state-metrics", label_alerts="channel", namespace="service", pod="prometheus-kube-state-metrics-748b59796-t4s29", service="prometheus-kube-state-metrics"}
```
3. and grouping the 2 metrics on `namespace` in order to get the value of the `alerts` label of the namespace from step 1. , and use it as a label on the alert itself
```
name: KubeJobFailed
expr: ( kube_job_status_failed{job="kube-state-metrics",namespace=~"{{ $targetNamespace }}"} > 0 ) * on(namespace) group_left(label_alerts) kube_namespace_labels
for: 1h
labels:
  severity: warning
  team: {{ $labels.label_alerts }}
annotations:
  message: Job {{ $labels.namespace }}/{{ $labels.job_name }} failed to complete.
  runbook_url: https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-kubejobfailed
```
4. this way we propagate the label from namespace to alerts, and it works as expected.

The problem is, that we need to copy/paste all the rules from the kube-state-metrics in order to add the grouping to the expressions `* on(namespace) group_left(label_alerts) kube_namespace_labels` and the parenthesis. I imagine others may face the same challenges, so I thought about adding 2 extra values like `.Values.defaultRules.expressionPrefix` and `.Values.defaultRules.expressionSuffix` to be able to feed the grouping into all the rules without copy/pasting and maintaining the rules ... so the rule would become something like:
```
    - alert: KubeJobFailed
      annotations:
        description: Job {{`{{`}} $labels.namespace {{`}}`}}/{{`{{`}} $labels.job_name {{`}}`}} failed to complete. Removing failed job after investigation should clear this alert.
        runbook_url: {{ .Values.defaultRules.runbookUrl }}alert-name-kubejobfailed
        summary: Job failed to complete.
      expr: {{ .Values.defaultRules.expressionPrefix }} kube_job_failed{job="kube-state-metrics", namespace=~"{{ $targetNamespace }}"}  > 0 {{ .Values.defaultRules.expressionSufix }}
      for: 15m
      labels:
        severity: warning
{{- if .Values.defaultRules.additionalRuleLabels }}
{{ toYaml .Values.defaultRules.additionalRuleLabels | indent 8 }}
{{- end }}
```
and the values:
```
defaultRules:
  expressionPrefix: "("
  expressionSufix: ") * on(namespace) group_left(label_alerts) kube_namespace_labels"
```
Let me know please WDYT, or if you have another alternative to achieve the same goal.

Cheers,
Robert

### PR DESCRIPTION
**Pull request by nrobert13**
Signed-off-by: Robert N <nrobert13@gmail.com>

#### What this PR does / why we need it:
allow alert routing in multi tenant clusters to different teams
#### Which issue this PR fixes
*(optional, in `fixes #<issue_number> (, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #1231 

#### Special notes for your reviewer:
hey @vsliouniaev , @bismarck, @gianrubio , @gkarthiks, @scottrigby, @Xtigyro, would you be so kind to review this change. Thanks in advance.
Cheers,
Robert
#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/kube-prometheus-stack/Chart.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ name: kube-prometheus-stack
 sources:
   - https://github.com/prometheus-community/helm-charts
   - https://github.com/prometheus-operator/kube-prometheus
-version: 18.0.0
+version: 18.0.2
```
Comment (monotek) [inline charts/kube-prometheus-stack/Chart.yaml:21]: adding new feature should raise chart minor version

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/kube-prometheus-stack/Chart.yaml (hunk 1 above)
Comment (nrobert13) [reply charts/kube-prometheus-stack/Chart.yaml]: @monotek , thanks for the feedback. updated the minor version.

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/general.rules.yaml [hunk 2]
```diff
@@ -29,7 +29,10 @@ spec:
         description: '{{`{{`}} printf "%.4g" $value {{`}}`}}% of the {{`{{`}} $labels.job {{`}}`}}/{{`{{`}} $labels.service {{`}}`}} targets in {{`{{`}} $labels.namespace {{`}}`}} namespace are down.'
         runbook_url: {{ .Values.defaultRules.runbookUrl }}alert-name-targetdown
         summary: One or more targets are unreachable.
-      expr: 100 * (count(up == 0) BY (job, namespace, service) / count(up) BY (job, namespace, service)) > 10
+      expr: |-
+        {{ .Values.defaultRules.expressionPrefix }}
```
Comment (jorik90) [inline charts/kube-prometheus-stack/templates/prometheus/rules-1.14/general.rules.yaml:33]: This is a change in a generated file (see the comment at the top of the file). When the next rules-sync is executed, your changes will be reverted.

**Review by jorik90 — CHANGES_REQUESTED** [review]
Two things worth noting:
1. You've made changes in file that are generated. These changes will be overwritten.
2. I've proposed an alternative solution to probably solve your problem. See my comment in https://github.com/prometheus-community/helm-charts/issues/1231#issuecomment-912613496

**Comment by nrobert13** [comment]
thanks for the proposed solution. closing this.

**Pull request closed without merging**